system/jlib/jtrace.cpp

```diff
@@ -69,3 +100,688 @@ const char* LogTrace::queryLocalId() const
 {
     return localId.get();
 }
+
+class CHPCCHttpTextMapCarrier : public opentelemetry::context::propagation::TextMapCarrier
+{
+public:
+    CHPCCHttpTextMapCarrier(const IProperties * httpHeaders)
+    {
+        if (httpHeaders)
+        {
+            this->httpHeaders.setown(createProperties());
+            Owned<IPropertyIterator> iter = httpHeaders->getIterator();
+            ForEach(*iter)
+            {
+                const char * key = iter->getPropKey();
+                const char * val = httpHeaders->queryProp(key);
+                this->httpHeaders->setProp(key, val);
+            }
+        }
+    }
+
+    CHPCCHttpTextMapCarrier()
+    {
+        httpHeaders.setown(createProperties());
+    };
+
+    virtual opentelemetry::nostd::string_view Get(opentelemetry::nostd::string_view key) const noexcept override
+    {
+        std::string theKey = key.data();
+
+        if (theKey.empty())
+            return "";
+
+        Owned<IPropertyIterator> iter = httpHeaders->getIterator();
+        ForEach(*iter)
+        {
+            const char * propKey = iter->getPropKey();
+            if (stricmp(propKey, theKey.c_str()) == 0)
+                return httpHeaders->queryProp(propKey);
+        }
+
+        return "";
+    }
+
+    virtual void Set(opentelemetry::nostd::string_view key, opentelemetry::nostd::string_view value) noexcept override
+    {
+        httpHeaders->setProp(std::string(key).c_str(), std::string(value).c_str());        
+    }
+
+private:
+    Owned<IProperties> httpHeaders;
+};
+
+enum SpanType { ServerSpan, ClientSpan, InternalSpan };
+
+class CSpan : public CInterfaceOf<ISpan>
+{
+public:
+    CSpan() : span(nullptr) {};
+    ~CSpan()
+    {
+        if (span != nullptr)
+        {
+            StringBuffer out;
+            toString(out);
+            DBGLOG("Span end: (%s)", out.str());
+            span->End();
+        }
+    }
+
+    ISpan * createClientSpan(const char * name) override
+    {
+        return new CSpan(name, SpanType::ClientSpan, this);
+    }
+
+    ISpan * createInternalSpan(const char * name) override
+    {
+        return new CSpan(name, SpanType::InternalSpan, this);
+    }
+
+    void toString(StringBuffer & out) const override
+    {
+        out.append("Type: ");
+        if (type == SpanType::ServerSpan)
+            out.append("Server");
+        else if (type == SpanType::ClientSpan)
+            out.append("Client");
+        else
+            out.append("Internal");
+
+        if (span != nullptr)
+        {
+            out.append(" Name: ").append(name.get())
+            .append(" SpanID: ").append(spanID.get())
+            .append(" TraceID: ").append(traceID.get())
+            .append(" TraceFlags: ").append(traceFlags.get())
+            .append(" HPCCGlobalID: ").append(hpccGlobalId.get())
+            .append(" HPCCCallerID: ").append(hpccCallerId.get());
+
+            if (localParentSpan != nullptr)
```

Comment:
localParentSpan == nested span's parent resides in same process
remoteParentSpan == server span's parent is declared via propagated context
Otel refers to these are parentspan, and parenttrace, i can refactor if we can think of better names